### PR TITLE
Support for dict comprehension

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

Closes: #5135

### CODE REVIEW

File: docs/source/reference/pysupported.rst [hunk 1]
```diff
@@ -796,6 +796,23 @@ range of possible failures. However, the dictionary can be safely read from
...
+Numba supports dictionary comprehension under the assumption that a
+``numba.typed.Dict`` instance can be created from the comprehension.  For
+example::
+
+  In [1]: from numba import njit
+
+  In [2]: @njit
+  ...: def foo(n):
+  ...:     return {i: i**2 for i in range(n)}
+  ...:
```
Comment (esc) [inline docs/source/reference/pysupported.rst:811]:
```suggestion
In [2]: @njit
   ...: def foo(n):
   ...:     return {i: i**2 for i in range(n)}
   ...:
```

File: numba/core/dataflow.py [hunk 2]
```diff
@@ -202,6 +202,16 @@ def op_BUILD_MAP(self, info, inst):
         info.append(inst, items=items[::-1], size=count, res=dct)
         info.push(dct)
 
+    def op_MAP_ADD(self, info, inst):
```
Comment (esc) [inline numba/core/dataflow.py:205]: I tried to check if this code was ever exercised, by using the patch:

```
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'vals' of function 'TestDictObject.test_dict_values.<locals>.diff --git i/numba/core/byteflow.py w/numba/core/byteflow.py
index 5c612f3a08..bf7f48cf84 100644
--- i/numba/core/byteflow.py
+++ w/numba/core/byteflow.py
@@ -920,6 +920,8 @@ class TraceRunner(object):
         # NOTE: https://docs.python.org/3/library/dis.html#opcode-MAP_ADD
         # Python >= 3.8: TOS and TOS1 are value and key respectively
         # Python < 3.8: TOS and TOS1 are key and value respectively
+        import ipdb
+        ipdb.set_trace()
         TOS = state.pop()
         TOS1 = state.pop()
         key, value = (TOS, TOS1) if PYVERSION < (3, 8) else (TOS1, TOS)
diff --git i/numba/core/dataflow.py w/numba/core/dataflow.py
index c351086a50..646103e682 100644
--- i/numba/core/dataflow.py
+++ w/numba/core/dataflow.py
@@ -203,6 +203,8 @@ class DataFlowAnalysis(object):
         info.push(dct)

     def op_MAP_ADD(self, info, inst):
+        import ipdb
+        ipdb.set_trace()
         key = info.pop()
         value = info.pop()
         index = inst.arg
```

But it never stopped in `datalflow.py` when executing the tests in this PR.

**Review by esc — CHANGES_REQUESTED** [review]
One suggestion to fix doc formatting and a question about test coverage.

File: numba/core/dataflow.py (hunk 2 above)
Comment (esc) [reply numba/core/dataflow.py:205]: So, `dataflow.py` is for older Pythons.

File: docs/source/reference/pysupported.rst (hunk 1 above)
Comment (stuartarchibald) [reply docs/source/reference/pysupported.rst]: well spotted, colons are lined up in https://github.com/numba/numba/pull/6736/commits/7145e6cdf6f9d53a93281d89c7efe66f871b3702

Comment (esc) [reply docs/source/reference/pysupported.rst]: excellent, ready to smoketest then?

Comment (stuartarchibald) [reply docs/source/reference/pysupported.rst]: Think so!

**Comment by esc** [comment]
Build farm ID: `numba_smoketest_cpu_91`

File: numba/core/dataflow.py (hunk 2 above)
Comment (sklam) [reply numba/core/dataflow.py:205]: Right, for python<3.7. I can't wait to drop py3.6.

**Review by esc — APPROVED** [review]

**Comment by esc** [comment]
Smoketest was green, this is good to go in.

**Pull request merged**